### ISSUE TITLE
hyper parameter optimization - killed for large num_leaves even though histogram_pool_size is set

### ISSUE THREAD

**Issue by geoHeil** [Microsoft/LightGBM#725]
I already had a question about hyper parameter optimization here https://github.com/Microsoft/LightGBM/issues/695 and https://github.com/Microsoft/LightGBM/issues/687 unfortunately, setting histogram_pool_size does not seem to help.

For example
```
assignments  Assignments({
  "bagging_fraction": 0.3,
  "colsample_bytree": 1.0,
  "max_depth": 100,
  "min_child_samples": 500,
  "min_child_weight": 5,
  "min_data_in_leaf": 500,
  "num_leaves": 4095,
  "scale_pos_weight": 0.01,
  "subsample": 0.4
})
```
result in `[1]    15668 killed     python3 myScript.py`

Now I was already setting the to 8GB. Trying again with 500 i.e. 500mb it fails again - but later. I also noticed that the memory is steadily increasing. Not sure if this is a fault of lightGBM though.

**Comment by guolinke** [comment]
Are you running many lightgbm instances ? 
If yes, you can try the our recently fix for the high memory consumption when running many instances.

https://github.com/Microsoft/LightGBM/commit/f7d190aa5c021de10fd5924d782d2cdf4eee9494 

In short, you can try the latest code of master branch.

**Comment by geoHeil** [comment]
Only a single one. Already in the latest master from last night.
Guolin Ke <notifications@github.com> schrieb am Di. 25. Juli 2017 um 06:52:

**Comment by guolinke** [comment]
@geoHeil 
How many iterations you are running ? 
And What is the total memories of you machines? 

It seems your are running the big tree model. So the memory will increase with the iterations since the model size is relative the num_leaves times num_iterations.

**Comment by geoHeil** [comment]
Default iterations i.e. 100. Memory is about 700GB. Possibly it is a combination of https://stackoverflow.com/questions/45287186/how-to-find-memory-leak-with-pandas as well.

**Comment by guolinke** [comment]
@geoHeil I think 100 iterations isn't a big deal.

Can you give a code to show how do you use lightgbm?

**Comment by geoHeil** [comment]
I use 
```
parameters = [
            dict(name="num_leaves", type="int", bounds=dict(min=7, max=4095)),
            dict(name="max_depth", type="int", bounds=dict(min=2, max=100)),
            dict(name="scale_pos_weight", type="double", bounds=dict(min=0.01, max=10000.0)),
            dict(name="min_data_in_leaf", type="int", bounds=dict(min=5, max=500)),
            dict(name="min_child_weight", type="int", bounds=dict(min=5, max=500)),
            dict(name="min_child_samples", type="int", bounds=dict(min=5, max=500)),
            dict(name="subsample", type="double", bounds=dict(min=0.4, max=1)),
            dict(name="bagging_fraction", type="double", bounds=dict(min=0.3, max=1)),
            dict(name="colsample_bytree", type="double", bounds=dict(min=0.4, max=1))
        ]
```
where these parameters are applied statically
        ```
static_parameters = {'boosting_type': 'dart', 'reg_alpha': 0, 'reg_lambda': 2, 'is_unbalance': False,
                             'min_split_gain': 0, 'subsample_freq': 3,
                             'subsample_for_bin': 50000,
                             'max_bin': 99999, "learning_rate": 0.1, "n_estimators": 9999999,
                             'histogram_pool_size': detect_available_memory_for_histogram_cache()}

def detect_available_memory_for_histogram_cache():
    res = psutil.virtual_memory().total
    small_memory_marker = 17179869184  # around 17GB in bytes (ram + swap)
    if res <= small_memory_marker:
        cache_size = 500  # 500 mb cache size
    else:
        cache_size = 51200  # 50gb

    logger.info("histogram_pool_size caching is " + str(cache_size))
    return cache_size
```
and e.g. a suggestion like:

```
assignments  Assignments({
  "bagging_fraction": 0.3,
  "colsample_bytree": 1.0,
  "max_depth": 100,
  "min_child_samples": 500,
  "min_child_weight": 5,
  "min_data_in_leaf": 500,
  "num_leaves": 4095,
  "scale_pos_weight": 0.01,
  "subsample": 0.4
})
```
is tried.

Then, lgbm is executed via:
```
estimator.fit(X_train, y_train,
                          early_stopping_rounds=30,
                          eval_set=[(X_test, y_test)],
                          eval_metric="binary_logloss",
                          verbose=False,
                          feature_name='auto',
                          categorical_feature='auto')
```
I think `categorical_feature` is still from the old api

**Comment by guolinke** [comment]
```max_bin = 99999``` ? This will use 4-times of memory. 
BTW, Did the training start (had outputed the metric if enable verbose) when you are seeing the increasing memory usage ?

**Comment by guolinke** [comment]
BTW, "subsample" and "bagging_fraction" are the same parameter.

**Comment by geoHeil** [comment]
@guolinke Thanks for the hint. that 99999 worked so far when not using hyperparameter optimization. Do you think that: 32767 is an useful value?

**Comment by guolinke** [comment]
@geoHeil it will reduce the half memory. 
BTW, will it be killed when running with small leaves ? can you test that ?

**Comment by geoHeil** [comment]
When choosing histogram_pool_size caching= 2048 and num_leves = 1024 does not seem to crash.

**Comment by guolinke** [comment]
@geoHeil 
I just notice you are using ``"n_estimators": 9999999`` .
So the memory will increase during training since you are train a very big model (9999999 * 4095) .

**Comment by geoHeil** [comment]
Yes, but as mentioned in one of the linked documents @Laurae2 suggests this behavior as early stopping prevents overfitting. Would you suggest a different approach?

**Comment by Laurae2** [comment]
Memory requirement is nearly quadratic to the number of leaves, if you are at the limit of your RAM at 2047 leaves (full 11 depth), it will explode RAM at 4095 leaves (full 12 depth).

@guolinke In the case of the number of iterations, if a validation dataset is available, it is better to set it to an extremely large value (such as 10000000) and let early stopping doing its job.

**Comment by GeorgianaPetria** [comment]
#geoHeil are you fine tuning with all those parameters? 
I'm getting `candidate_params = list(self._get_param_iterator()) MemoryError` if I run GridSearchCV with only some of them...

**Comment by geoHeil** [comment]
Now with the fixes of some memory leaks and manually setting the allowed
memory it works fine

GeorgianaPetria <notifications@github.com> schrieb am Sa. 21. Okt. 2017 um
22:22:

**Comment by GeorgianaPetria** [comment]
I'm not sure I understood. I am still getting a `MemoryError`.
Could you please summaries what parameters you are optimising?

Also, I couldn't use the exact structure you are using for the parameters because it complained GridSearchCV and LGBMClassifier require a sequence of values.
`parameters = [
            dict(name="num_leaves", type="int", bounds=dict(min=7, max=4095)),
           ...
        ]
 `
Where are you passing this structure?

**Comment by geoHeil** [comment]
That notation of parameters is not applicable when using grid search.

It can be used though to specify a parameter grid for sigOpt a commercial
parameter tuning service.

By the way I would recommend random instead of grid search if you rely on
using sklearn builtins.
https://medium.com/rants-on-machine-learning/smarter-parameter-sweeps-or-why-grid-search-is-plain-stupid-c17d97a0e881
GeorgianaPetria <notifications@github.com> schrieb am So. 22. Okt. 2017 um
21:10:

**Comment by guolinke** [comment]
@GeorgianaPetria you should limit ```histogram_pool_size``` if your machine's Memory is not enough.
